### ISSUE TITLE
🦠 Model Request: Tox21 p54

### ISSUE THREAD

**Issue by GemmaTuron** [ersilia-os/ersilia#463]
### Model Title

Tox21 p53

### Publication

Hello @Ng-ethe !

As part of your Outreachy contribution, we have assigned you the dataset "Tox21-p54" from the Therapeutics Data Commons to try and build a binary classification ML model.
Please copy the provided Google Colab template and use this issue to provide updates on the progress. We'll value not only being able to build the model but also interpreting its results.

**Comment by Ng-ethe** [comment]
Great! On it.

**Comment by Ng-ethe** [comment]
[](https://colab.research.google.com/drive/1nZTEfEIYeCV7ytLstp1PkoKRybfk-9FY?usp=sharing)

This is the link to the Colab file I'm using.

**Comment by Ng-ethe** [comment]
Also, I interpreted this as `Tox21- SR-p54`. From the listed labels, only `SR-p53` is present but passing `SR-p54` to the `label_name` parameter also works so I went with that.

**Comment by Ng-ethe** [comment]
The first step was checking out the [Therapeutics Data Commons](https://tdcommons.ai/) website. They give instructions and guidelines for installing required packages, how to use functions and how to access datasets through the provided APIs.

On google Colab (using the provided `LazyQSAR` notebook template, I installed Therapeutics Data Commons by running `!pip install PyTDC` and then importing tdc using `import tdc`. Tox21 datase is a Single-instance Prediction problem. It is a data challenge which contains qualitative toxicity measurements for 7,831 compounds on 12 different targets, such as nuclear receptors (denoted as NR) and (stree response pathways denoted as SR). The following are the 12 targets which can be loaded using `from tdc.utils import retrieve_label_name_list
label_list = retrieve_label_name_list('Tox21')
label_list`

1. NR-AR
2.  NR-AR-LBD
3.  NR-AhR
4.  NR-Aromatase
5.  NR-ER
6.  NR-ER-LBD
7.  NR-PPAR-gamma
8.  SR-ARE
9.  SR-ATAD5
10.  SR-HSE
11.  SR-MMP
12.  SR-p53

The specific target is `SR-p54` and can be loaded as follows

 `from tdc.single_pred import Tox
data = Tox(name = 'Tox21', label_name = 'SR-P54')
df = data.get_data()
df.sample(25)`

**Comment by Ng-ethe** [comment]
![image](https://user-images.githubusercontent.com/58877986/198881251-aa1e830b-2101-4086-9a9a-70a931bb558f.png)

This is a preview of the dataset.

**Comment by Ng-ethe** [comment]
![image](https://user-images.githubusercontent.com/58877986/198881308-5d743421-6d1a-4cfb-ab5c-6d924689740e.png)

Then the data has to be split into train, test and validation sets. The method used for splitting datasets by default is the random method but for Tox21, the scaffold method can also be used Scaffold split which only applies to single-instance drugs-related tasks (ADME, Tox, HTS) and requires RDKit installed.  It is based on the scaffold of the molecules so that train, validation and test sets are more structurally different. However, itt is more challenging than the random split.

**Comment by Ng-ethe** [comment]
The dataset has 6773 entries. 
The train dataset has 4741 molecules, 4512 inactive and 229 active.
The test dataset has 1356 molecules, 142 active and 1214 inactive.
The validation dataset has 677 molecules, 52 active and 625 inactive.

**Comment by Ng-ethe** [comment]
![image](https://user-images.githubusercontent.com/58877986/198913583-27451b2a-48d0-4a64-b6d0-94cbf3134d8c.png)
![image](https://user-images.githubusercontent.com/58877986/198913589-aca9a3fa-f4b5-4e16-9e6c-53ceb82ed5af.png)
![image](https://user-images.githubusercontent.com/58877986/198913605-78374089-eec8-451f-ba5f-2bd273de66c6.png)


these are the visualizations for the datasets using matplotlib.

**Comment by Ng-ethe** [comment]
![image](https://user-images.githubusercontent.com/58877986/198913850-86aa48de-2df0-40b7-bf85-5280df2a5b9c.png)
 

Using `rdkit `to draw an active molecule. The second image shows the same molecule but with the atom indices included.

**Comment by Ng-ethe** [comment]
![image](https://user-images.githubusercontent.com/58877986/198913981-1411b7f5-373f-42b8-bd6e-9d1cebea1876.png)
 
Using rdkit and IPython console to draw an inactive molecule.

**Comment by Ng-ethe** [comment]
I then split the 3 datasets into attributes and labels
I used the command `!pip install git+https://github.com/ersilia-os/lazy-qsar.git` to install Lazy-QSAR on Google Colab. Since Tox21 is a Binary Classification problem the Morgan Binary Classifier was used. A class of the classifier was created and fit with the training data. SciKitLearn was then used to make predictions and evaluate the model. Both the test and validation models were used for the last step.

**Comment by Ng-ethe** [comment]
These are the ROC curves obtained.

![image](https://user-images.githubusercontent.com/58877986/198914653-f09de2e5-5196-4ccf-aca9-b5d0e67787c7.png)
![image](https://user-images.githubusercontent.com/58877986/198914707-5c721f8e-677f-4624-963a-f0030f6b84f7.png)
Both ROC curves have an AUC greater than 0.5, meaning they rank a random positive example higher than a random negative example more than 50% of the time. Real-world binary classification AUC values generally fall into this range.

**Comment by Ng-ethe** [comment]
In conclusion, the Tox21 - SR-p54 TDC dataset was loaded and analyzed. The dataset has a class imbalance since inactive molecules are more than the active molecules. The data was modeled using the Morgan Binary Classifier since this task was a binary classification problem. The model was evaluated using SciKit-Learn using the accuracy, precision, recall, F1 score, ROC and AUC metrics. The model performance is shown in the table below.

|           | Test       | Validation |
|-----------|------------|------------|
|Accuracy   | 0.895      | 0.923
|Precision  | 0.895      | 0.923
|Recall     | 0.895      | 0.923
|F1 Score   | 0.895      | 0.923
|AUROC      | 0.621      | 0.564

**Comment by carcablop** [comment]
Hello @Ng-ethe  
I have noticed that your metrics are very good even though the data is not balanced, there is a great imbalance, since I see that your inactive molecules are much more than the active molecules.
In this case, are you calculating the precision and recall for the case in which the molecules are inactive? In other words, in your formula are you taking into account only the negative ones? Are you interested in non-toxic molecules? I ask the question because I used a tox21 data set, where I also had unbalanced data, and what really matters are the cases in which the molecules are toxic
For example, measuring the accuarancy in this model is not useful, and it could cause us confusion by believing that we have a model that is doing the predictions well but it turns out that it is doing it well but for the cases that are negative or in this case for the molecules that they are not toxic.
In this case I would expect your recall metrics to be very low. Also your AUROC is 60%, it makes more sense for a model where the data is not balanced.
It would also be good if you shared your confusion matrix.

**Comment by Ng-ethe** [comment]
Thank you @carcablop. I definitely thought about that and had the same concern. The most sensible thing I can up with is getting the recall and precision for each of the two classes by setting `average=None`. As seen below
![image](https://user-images.githubusercontent.com/58877986/198923916-df200b21-e482-496b-8d65-1c1fe2961012.png)
![image](https://user-images.githubusercontent.com/58877986/198924053-f95a9cac-f7e1-4a2f-9298-a5afce7b9175.png)

 The scores for the active/toxic molecules are nearly or equal to zero.

**Comment by GemmaTuron** [comment]
Hi  @Ng-ethe,

I am concerned about your results seeing how your ROC curve is actually not very good. I have seen you are using the predict function instead of the predict_proba and taking the probability of 1 for further analysis using contingency tables and precision and recall.

**Comment by carcablop** [comment]
Hello @Ng-ethe.
I have not seen your colab template, my previous comment was because I saw your precision and recall metrics and the results seemed very strange to me, I hope I have not confused you.
The first thing you should do is fix the problem you are having with the roc curve.
And if I also see that you are not using the function well, you should use the **predict_prob** function to obtain more accurate data in the evaluation when the model is evaluated, you will see that your roc curve will improve, that is, the value of your AUROC should be closer to 1 , remember that 1 means that the predictions are 100% correct. Remember that these results are probabilities, and you must then convert them to zeros and ones to use in your confusion matrix.

**Comment by Ng-ethe** [comment]
> Hi @Ng-ethe,
> 
> I am concerned about your results seeing how your ROC curve is actually not very good. I have seen you are using the predict function instead of the predict_proba and taking the probability of 1 for further analysis using contingency tables and precision and recall.

@GemmaTuron I used `predict` for calculating the precision, recall and accuracy because using predictions from `predict_proba`  raises the shown error

![image](https://user-images.githubusercontent.com/58877986/199148577-893e19da-6d37-48fc-9df9-dd1f8fd2d3ba.png)

**Comment by Ng-ethe** [comment]
> 

These are the new ROC curves I obtained, with an AUC of 0.75 and 0.73.

![image](https://user-images.githubusercontent.com/58877986/199148791-62a4d7e1-9e8d-4d61-8649-9870af9e5967.png)

![image](https://user-images.githubusercontent.com/58877986/199148878-81c2bf14-1b35-4024-8801-e52832c35922.png)

**Comment by Ng-ethe** [comment]
> Hello, @Ng-ethe. I have not seen your colab template, my previous comment was because I saw your precision and recall metrics and the results seemed very strange to me, I hope I have not confused you. The first thing you should do is fix the problem you are having with the roc curve. And if I also see that you are not using the function well, you should use the **predict_prob** function to obtain more accurate data in the evaluation when the model is evaluated, you will see that your roc curve will improve, that is, the value of your AUROC should be closer to 1, remember that 1 means that the predictions are 100% correct. Remember that these results are probabilities, and you must then convert them to zeros and ones to use in your confusion matrix.

Hello @carcablop, I used the `predict` function to get y_pred values to use with the accuracy, precision and the recall evaluation metrics. However, I did use `predict_probab`. I ran the code again and the curve looks much better with an AUC value of .75 and .73 for the test and validation dataset.


My question: should I also use `predict_proba` and convert the results to ones and zeros so as to calculate accuracy, precision and recall? Or is `predict` just okay for these specific cases?

**Comment by Ng-ethe** [comment]
> > 
> 
> These are the new ROC curves I obtained, with an AUC of 0.75 and 0.73.
> 
> ![image](https://user-images.githubusercontent.com/58877986/199148791-62a4d7e1-9e8d-4d61-8649-9870af9e5967.png)
> 
> ![image](https://user-images.githubusercontent.com/58877986/199148878-81c2bf14-1b35-4024-8801-e52832c35922.png)

@carcablop here are the new curves I got?

**Comment by carcablop** [comment]
> > Hello, @Ng-ethe. I have not seen your colab template, my previous comment was because I saw your precision and recall metrics and the results seemed very strange to me, I hope I have not confused you. The first thing you should do is fix the problem you are having with the roc curve. And if I also see that you are not using the function well, you should use the **predict_prob** function to obtain more accurate data in the evaluation when the model is evaluated, you will see that your roc curve will improve, that is, the value of your AUROC should be closer to 1, remember that 1 means that the predictions are 100% correct. Remember that these results are probabilities, and you must then convert them to zeros and ones to use in your confusion matrix.
> 
> Hello @carcablop, I used the `predict` function to get y_pred values to use with the accuracy, precision and the recall evaluation metrics. However, I did use `predict_probab`. I ran the code again and the curve looks much better with an AUC value of .75 and .73 for the test and validation dataset.
> 
> My question: should I also use `predict_proba` and convert the results to ones and zeros so as to calculate accuracy, precision and recall? Or is `predict` just okay for these specific cases?

Hello @Ng-ethe. 
Of course, you should also use the `predict_prob `function, these results, depending on the threshold, will be converted to 0 or 1, and they are the ones that you pass to the confusion matrix, and then you use it to make the calculations of the metrics such as precision and recall.
An example to calculate the confusion matrix is the following:

f```
rom sklearn.metrics import confusion_matrix
from sklearn.metrics import ConfusionMatrixDisplay

confusion_matrix_valid= confusion_matrix(Y_val, new_Y_prediction_val)
#"new_Y_prediction_val: is the list you get from having applied the threshold.

cfmatrdisp= ConfusionMatrixDisplay(confusion_matrix=confusion_matrix_valid)
cfmatrdisp.plot()
```

When you get the results of the confusion matrix, and calculate the metrics by hand, you will see everything more clearly, and if your model is really good at classifying and how you could improve it.

**Comment by Ng-ethe** [comment]
@carcablop thank you for the feedback. I'll try implement that.

**Comment by GemmaTuron** [comment]
Hello @Ng-ethe  !

Thanks for your contributions to Ersilia during the Outreachy Application Period! We hope you have learnt and enjoyed as much as the Ersilia team did!
I will close this issue as this was part of the application period solely.
